File: Nerava/Nerava/Support/DeepLinkHandler.swift
import Foundation

enum DeepLinkHandler {
    /// Resolves incoming URL to web app URL for loading in WebView
    /// Supports:
    /// - Universal Links: https://app.nerava.network/...
    /// - Custom scheme: nerava://merchant/{id}, nerava://session/{id}
    static func resolveWebURL(from url: URL) -> URL? {
        // Universal Link - use as-is if it's our domain
        if url.scheme == "https" && isOurDomain(url.host) {
            return url
        }

        // Custom scheme: nerava://
        if url.scheme == "nerava" {
            return resolveCustomScheme(url)
        }

        return nil
    }

    private static func isOurDomain(_ host: String?) -> Bool {
        guard let host = host else { return false }
        return host == "app.nerava.network" || host == "link.nerava.network"
    }

    private static func resolveCustomScheme(_ url: URL) -> URL? {
        let baseURL = Environment.current.webAppURL
        let path = url.path

        // nerava://merchant/123 -> https://app.nerava.network/merchant/123
        // nerava://session/abc -> https://app.nerava.network/session/abc
        // nerava://exclusive/xyz -> https://app.nerava.network/exclusive/xyz
        if !path.isEmpty {
            return baseURL.appendingPathComponent(path)
        }

        // Fallback to base URL
        return baseURL
    }

    /// Supported deep link paths
    static let supportedPaths = [
        "/merchant/",      // Merchant detail: /merchant/{id}
        "/session/",       // Active session: /session/{id}
        "/exclusive/",     // Exclusive detail: /exclusive/{id}
        "/s/",             // Short link: /s/{code}
        "/m/"              // Short merchant link: /m/{merchantId}
    ]
}

File: mobile/nerava_android/app/src/main/java/network/nerava/app/deeplink/DeepLinkHandler.kt
package network.nerava.app.deeplink

import android.content.Intent
import android.net.Uri
import android.util.Log

/**
 * Handles deep links and Android App Links.
 *
 * Supported schemes:
 * - nerava://merchant/{id} → loads web app at /merchant/{id}
 * - nerava://session/{id} → loads web app at /session/{id}
 * - https://app.nerava.network/{path} → loads directly in WebView
 *
 * Note: iOS does NOT implement deep links yet. This is Android-first but
 * follows the same URL patterns the web app router already handles.
 */
object DeepLinkHandler {

    private const val TAG = "DeepLinkHandler"

    /**
     * Extract a web URL from an incoming deep link intent.
     * Returns null if the intent doesn't contain a valid deep link.
     */
    fun resolveWebUrl(intent: Intent?, webAppBaseUrl: String): String? {
        val uri = intent?.data ?: return null
        return resolveUri(uri, webAppBaseUrl)
    }

    fun resolveUri(uri: Uri, webAppBaseUrl: String): String? {
        Log.i(TAG, "Resolving deep link: $uri")

        return when (uri.scheme) {
            "nerava" -> resolveCustomScheme(uri, webAppBaseUrl)
            "https" -> resolveAppLink(uri, webAppBaseUrl)
            else -> {
                Log.w(TAG, "Unknown scheme: ${uri.scheme}")
                null
            }
        }
    }

    private fun resolveCustomScheme(uri: Uri, webAppBaseUrl: String): String? {
        val path = uri.pathSegments
        if (path.isEmpty()) return webAppBaseUrl

        return when (path[0]) {
            "merchant" -> {
                val merchantId = path.getOrNull(1) ?: return webAppBaseUrl
                "$webAppBaseUrl/merchant/$merchantId"
            }
            "session" -> {
                val sessionId = path.getOrNull(1) ?: return webAppBaseUrl
                "$webAppBaseUrl/session/$sessionId"
            }
            "exclusive" -> {
                val exclusiveId = path.getOrNull(1) ?: return webAppBaseUrl
                "$webAppBaseUrl/exclusive/$exclusiveId"
            }
            else -> {
                // Pass through: nerava://path → webAppBaseUrl/path
                "$webAppBaseUrl/${path.joinToString("/")}"
            }
        }
    }

    private fun resolveAppLink(uri: Uri, webAppBaseUrl: String): String? {
        val host = uri.host
        if (host != "app.nerava.network") {
            Log.w(TAG, "App link from unexpected host: $host")
            return null
        }

        // The URL is already the web app URL — load it directly
        return uri.toString()
    }
}
